promql/engine_test.go

```diff
@@ -3096,6 +3101,146 @@ func TestInstantQueryWithRangeVectorSelector(t *testing.T) {
+	engine := newTestEngine()
+
+	baseT := timestamp.Time(0)
+	storage := promqltest.LoadedStorage(t, `
+		load 1m
+			some_metric{env="1"} 1
+			some_other_metric{env="1"} 2
+	`)
+	t.Cleanup(func() { require.NoError(t, storage.Close()) })
+
+	testCases := map[string]struct {
+		expr        string
+		expected    promql.Vector
+		expectedErr string
+		ts          time.Time
+	}{
+		"does not drop __name__ for instant queries": {
+			expr: "{__name__!=\"\"}",
+			ts:   baseT,
+			expected: promql.Vector{
+				promql.Sample{
+					F:      1,
+					T:      0,
+					Metric: labels.FromStrings("__name__", "some_metric", "env", "1"),
+				},
+				promql.Sample{
+					F:      2,
+					T:      0,
+					Metric: labels.FromStrings("__name__", "some_other_metric", "env", "1"),
+				},
+			},
+		},
+		"drops __name__ for functions": {
+			expr:        "count_over_time({__name__!=\"\"}[1m])",
+			ts:          baseT,
+			expectedErr: "vector cannot contain metrics with the same labelset",
+		},
+		"drops __name__ for unary expressions": {
+			expr:        "-{__name__!=\"\"}",
+			ts:          baseT,
+			expectedErr: "vector cannot contain metrics with the same labelset",
+		},
+		"drops __name__ for binary operations": {
+			expr:        "{__name__!=\"\"} + {__name__!=\"\"}",
+			ts:          baseT,
+			expectedErr: "vector cannot contain metrics with the same labelset",
+		},
+		"drops __name__ for vector-scalar binary operations": {
+			expr:        "{__name__!=\"\"} * 2",
+			ts:          baseT,
+			expectedErr: "vector cannot contain metrics with the same labelset",
+		},
+		"allows relabeling using __name__ via label_replace": {
+			expr: "label_replace(count_over_time({__name__!=\"\"}[1m]), \"original_name\", \"$1\", \"__name__\", \"(.+)\")",
+			ts:   baseT,
+			expected: promql.Vector{
+				promql.Sample{
+					F:              1,
+					T:              0,
+					Metric:         labels.FromStrings("original_name", "some_metric", "env", "1"),
+					ShouldDropName: true,
+				},
+				promql.Sample{
+					F:              1,
+					T:              0,
+					Metric:         labels.FromStrings("original_name", "some_other_metric", "env", "1"),
+					ShouldDropName: true,
+				},
+			},
+		},
+		"allows relabeling using __name__ via label_join": {
+			expr: "label_join(count_over_time({__name__!=\"\"}[1m]), \"original_name\", \"-\", \"__name__\", \"env\")",
+			ts:   baseT,
+			expected: promql.Vector{
+				promql.Sample{
+					F:              1,
+					T:              0,
+					Metric:         labels.FromStrings("original_name", "some_metric-1", "env", "1"),
+					ShouldDropName: true,
+				},
+				promql.Sample{
+					F:              1,
+					T:              0,
+					Metric:         labels.FromStrings("original_name", "some_other_metric-1", "env", "1"),
+					ShouldDropName: true,
+				},
+			},
+		},
+		"allows preserving __name__ via label_replace": {
+			expr: "label_replace(count_over_time({__name__!=\"\"}[1m]), \"__name__\", \"$1\", \"__name__\", \"(.+)\")",
```

Comment:
I used this because it was easier to test (run into some difficulties using the `rate` function in tests), but I will give it another try, as I agree it makes sense to test the expected use case